src/instance.rs

```diff
@@ -119,15 +122,47 @@ impl<T> Py<T> {
     }
 }
 
-pub trait AsPyRef<T: PyTypeInfo>: Sized {
+/// Retrives `&'py` types from `Py<T>` or `PyObject`.
+///
+/// # Examples
+/// `PyObject::as_ref` returns `&PyAny`.
+/// ```
+/// # use pyo3::prelude::*;
+/// use pyo3::ObjectProtocol;
+/// let obj: PyObject = {
+///     let gil = Python::acquire_gil();
+///     let py = gil.python();
+///     py.eval("[]", None, None).unwrap().to_object(py)
+/// };
+/// let gil = Python::acquire_gil();
+/// let py = gil.python();
+/// assert_eq!(obj.as_ref(py).len().unwrap(), 0);  // PyAny implements ObjectProtocol
+/// ```
+/// `Py<T>::as_ref` returns `&PyDict`, `&PyList` or so for native types, and `&PyCell<T>`
+/// for `#[pyclass]`.
+/// ```
+/// # use pyo3::prelude::*;
+/// use pyo3::ObjectProtocol;
+/// let obj: PyObject = {
+///     let gil = Python::acquire_gil();
+///     let py = gil.python();
+///     py.eval("[]", None, None).unwrap().to_object(py)
+/// };
+/// let gil = Python::acquire_gil();
+/// let py = gil.python();
+/// assert_eq!(obj.as_ref(py).len().unwrap(), 0);  // PyAny implements ObjectProtocol
+/// ```
+pub trait AsPyRef: Sized {
```

Comment:
:+1:

Reply:
Hmm but since
- `AsPyRef` is just for `Py<T>` and `PyObject`(and PyObject is actually Py<PyAny>)
- It requires lots of changes
I decided to leave it as is

Reply:
I suppose `AsPyRef` will need further changes after #687 anyway